pkg/flow/deduper.go

```diff
@@ -0,0 +1,100 @@
+package flow
+
+import (
+	"container/list"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+var dlog = logrus.WithField("component", "flow/Deduper")
+var timeNow = time.Now
+
+// deduperCache implement a LRU cache whose elements are evicted if they haven't been accessed
+// during the expire duration.
+// It is not safe for concurrent access.
+type deduperCache struct {
+	expire time.Duration
+	// key: RecordKey with the interface and MACs erased, to detect duplicates
+	// value: listElement pointing to a struct entry
+	ifaces map[RecordKey]*list.Element
+	// element: entry structs of the ifaces map ordered by expiry time
+	ll *list.List
+}
+
+type entry struct {
+	key        *RecordKey
+	ifIndex    uint32
+	expiryTime time.Time
+}
+
+// Dedupe receives flows and filters these belonging to duplicate interfaces. It will forward
+// the flows from the first interface coming to it, until that flow expires in the cache
+// (no activity for it during the expiration time)
+func Dedupe(expireTime time.Duration) func(in <-chan []*Record, out chan<- []*Record) {
+	cache := &deduperCache{
+		expire: expireTime,
+		ll:     list.New(),
+		ifaces: map[RecordKey]*list.Element{},
+	}
+	return func(in <-chan []*Record, out chan<- []*Record) {
+		for records := range in {
+			cache.removeExpired()
+			fwd := make([]*Record, 0, len(records))
+			for _, record := range records {
+				if !cache.isDupe(&record.RecordKey) {
+					fwd = append(fwd, record)
+				}
```

Comment:
Possible improvement: adding a metric here to count the number of duplicate.

Reply:
I added a debug log below. If you mean prometheus metrics, maybe we can create a new task to integrate a metrics server here.